### ISSUE TITLE
Feature: Astro Expressive Code should handle Astro's `<Code />` component

### ISSUE THREAD

**Issue by hirasso** [expressive-code/expressive-code#28]
Not sure if in the right repo for my request – but here we go 😄

I have `astro-expressive-code` installed on my [Astro](https://astro.build/) site. I'm also making use of astro's built-in [`<Code />`](https://docs.astro.build/en/reference/api-reference/#code-) component for some code that I'm reading during build time. Something along the lines of:

```astro
---
import { Code } from "astro/components";
import { readJsSourceFile } from "../js/backend.js";
const sourceCode = readJsSourceFile();
---
<Code code={sourceCode} lang="js" />
```

I noticed that `astro-expressive-code` doesn't pick up the component. Makes kind of sense, since that component doesn't (yet?) support the features provided by expressive-code (title, marking, ...)

Is there an integration planned for this?

**Comment by hirasso** [comment]
...The reason I'm kind of forced to use the component is because I haven't found a way to expand the variable inside a fenced code block. This:
~~~
```js
{sourceCode}
```
~~~
will simply print 
```js
{sourceCode}`
```
...instead of the actual source code

**Comment by hippotastic** [comment]
Ha, yet another totally valid use case I didn't think of yet! I really love all the input! 😄

I just checked the [source code](https://github.com/withastro/astro/blob/main/packages/astro/components/Code.astro) of the Astro `Code` component and it doesn't seem to provide a way of hooking into it. However, I could imagine exporting our very own `Code` component from the `astro-expressive-code` package as a drop-in replacement. You would then only need to change the `import` statement and keep using `<Code>` as is in your documents.

The only thing I'm not sure of is if we could access the Astro config in some way from that component so that the `Code` component could automatically use the site-wide settings you've provided in the config.

**Comment by hirasso** [comment]
So nice that you are so open for new ideas! ❤️

I'm afraid that keeping a drop-in `<Code />` component in sync with astro's implementation could become quite a chore down the road.

In my opinion, the ideal solution would be to gain first-party support for expressive-code features in the built-in Astro component. Being able to do something like this would be amazing:

```
<Code code={JS} lang="js" ec-title="script.js" ec-mark={["substring1", "substring2"]} ec-group:start />
```

Alternatively, an official second component would maybe be an even better approach:

```
<ExpressiveCode code={JS} lang="js" title="script.js" mark={["substring1", "substring2"]} group:start />
```

Do you think we could reach out to the Astro team for that? To be honest, I actually thought this project was tightly coupled with Astro already, since I found out about it through their docs 😅

**Comment by hippotastic** [comment]
> So nice that you are so open for new ideas! ❤️

Thank you for the kind words! I really appreciate all the great input.

> I'm afraid that keeping a drop-in `<Code />` component in sync with astro's implementation could become quite a chore down the road.

Hmm, I'm not sure if I understand why this could be difficult. My idea here was: As long as the Expressive Code version of this component has the same props as Astro's component, migrating any of your pages to the Expressive Code version should be as easy as editing the frontmatter section of your MDX document and replacing the line `import { Code } from 'astro/components';` with `import { Code } from 'astro-expressive-code';`.

The idea to export an alternate `<Code>` component (which could also be called differently of course) from `astro-expressive-code` came from looking at the existing [`<Prism>` component](https://docs.astro.build/en/reference/api-reference/#prism-) which is also provided by an integration. I was hoping that if this works for Prism, it should work for Expressive Code, too. At least I'd be willing to give it a good try!

> Do you think we could reach out to the Astro team for that? To be honest, I actually thought this project was tightly coupled with Astro already, since I found out about it through their docs 😅

I'm actually an Astro maintainer and Expressive Code is a refactored version of my code that I added to Astro Docs a long time ago to improve the code examples in the official documentation. 😄 We're now using Expressive Code in Astro Docs.

I have already asked my team at Astro about how I could implement a Code component that has access to the central Expressive Code configuration in your site settings (which is quite a challenge due to the way the Astro config is loaded and abstracted away from the actual pages, especially in SSR mode). There seems to be a solution which would require creating a separate config file like `expressive-code.config.mjs` and making it available to components through Vite's virtual modules feature. I can't promise how quickly I would able to try this, though!

**Comment by hippotastic** [comment]
Oh, and regarding having Expressive Code support in the built-in Astro component: I don't think that this would be a preferred option for the Astro team as they always focus on keeping their core package small and performant (and rightfully so in my opinion). They would most likely prefer to keep such functionality in a separate package.

**Comment by hirasso** [comment]
> Hmm, I'm not sure if I understand why this could be difficult. 

That was a misunderstanding – I didn't mean difficult for me, I meant difficult for you as the maintainer 😄

I'd love to see the feature, for me as an end user it would of course make no difference if I would import the component from `expressive-code` instead.

Also, thanks so much for the detailed information! Very interesting to learn more about the deeper considerations in Astro's core team.

Actually, I have a question unrelated to this issue – could we get in contact directly? My mail: mail@rassohilber.com

**Comment by hippotastic** [comment]
Actually that wasn't a misunderstanding! I'm not seeing much of a difficulty to maintain this right now. The Code component's public API seems to be fairly stable judging from its commit history. Famous last words maybe! 😆

So, I'll leave this very valid feature idea open to put it on the backlog! If anyone wants to give it a shot, feel free to let us know here! Otherwise I'll implement it once I have finished some other work, namely a new inline comment parser and the grouped tabs we discussed in another thread.

Sent you an email!

**Comment by hirasso** [comment]
I started [working on this](https://github.com/expressive-code/expressive-code/pull/32), but quickly hit a speed bump. When I run `build`, I'm getting this error message: 

> No loader is configured for ".astro" files: src/components/Code.astro

Which I guess originates from me attempting to [re-export the `.astro` file from `index.ts`](https://github.com/hirasso/expressive-code/blob/05deb2107a084da83e0bb7654a34c591f716858b/packages/astro-expressive-code/src/index.ts#L5):

```
export { default as Code } from './components/Code.astro'
```

Can you point me in the right direction here? 🙏

**Comment by hippotastic** [comment]
> I started [working on this](https://github.com/expressive-code/expressive-code/pull/32) and already hit a speed bump 😄
> 
> When I run `build`, I'm getting this error message:
> 
> > No loader is configured for ".astro" files: src/components/Code.astro
> 
> Can you point me in the right direction here? 🙏

First of all, thank you for jumping into this! I had a look at the code and I totally understand why you got stuck. The issue seems to be that it's actually hard to properly set up a package that both has a TS-based integration that needs to go through a transpiler (which is done by the build step that I set up in this package), but also exports some components that must not be transpiled but exported as-is (like `.astro` components).

I have asked the Astro team if they can provide guidance on how to do this the right way, and I hope that they can set up a small example or template for this. The other helpful templates they have are focused on component-only packages, and the tricky combination with a TypeScript integration isn't covered yet. The team told me that we're in very good company and that many developers ran into this exact problem during a recent hackathon, so they are aware of this and will provide official help.

I'd therefore probably advise to put this on hold until we have a good minimal example on how to do it properly. If you don't want to wait, I was pointed at the [`@astrojs/image`](https://github.com/withastro/astro/tree/main/packages/integrations/image/) integration, which apparently features this combination, but there's a lot of stuff going on in there, so it would probably take some time to dig through and extract the changes we'd need to make to get this to work.

**Comment by hirasso** [comment]
Thanks @hippotastic for your help with this! No rush from my end 🏝️😊 My [current main project](https://github.com/swup/fragment-plugin) is eating enough time already anyways 😁

**Comment by fviolette** [comment]
A big +1 for this, especially for the ability to inclure variables 💯

**Comment by nerdymomocat** [comment]
A big +1 for this. I have been working on [webtrotion](https://github.com/nerdymomocat-templates/webtrotion-astro-notion-cms-website-blog), which doesn't use markdown but rather renders notion's json directly into astro components. Having a ECode component there would be so awesome to allow for comment highlights or folding.

**Comment by hippotastic** [comment]
Thank you for your input to this! I also needed a similar component myself for EC's documentation website, where I had to come up with a quick way to render all available themes without their CSS being present on all other pages as well.

If you know your config, it's not hard to create such a component manually and use it in your project:

https://github.com/expressive-code/expressive-code/blob/main/docs/src/components/ThemePreview.astro

The tricky part in generalizing this for all Astro EC users is the "know your config" part as during rendering, we don't have access to Astro's config file and functions / instances created there. So I'll need to create the possibility to place an EC config file somewhere in your project which can then be detected and used automatically both from the initialization function in the Astro config file and an EC Code component.

This feature request is probably the one I'll work on next.

**Comment by nerdymomocat** [comment]
Thank you for pointing me to this! I would know my config, so I do not think that would be a problem. I was wondering if you have suggestions about best practices -- specifically, should the `new ExpressiveCode` be called only once through a whole build process? I know that Astro mentions to use their `cached shiki highlighter`. Is there something similar for expressive code too?

**Comment by hippotastic** [comment]
I also recommend reusing a single instance of the engine throughout the build process:

https://expressive-code.com/reference/core-api/#constructors

But that being said, I'd still advise you to try it without any optimizations first and only start doing more when you actually notice any performance issues.

**Comment by hippotastic** [comment]
I just released version 0.32.0 of Expressive Code to NPM, which adds a `<Code>` component to `astro-expressive-code`: https://expressive-code.com/key-features/code-component/

I'll now create a PR to add support for this new component to Starlight as well. Some minor modifications to the Starlight EC integration are required to make it work, but it shouldn't take long. :)

**Comment by hirasso** [comment]
🤩 That's amazing news! Also, congrats to the (new?) docs site!

**Comment by nerdymomocat** [comment]
This is awesome! I will look into this over the course of the month. Thank you ❤️
